### PR TITLE
MDEV-7611 mysqldump --dump-slave always starts stopped slave

### PR DESCRIPTION
**Pull request by ParadoxV5**
* [x] *The Jira issue number for this PR is: [MDEV-7611][]*
  * This bug’s reported 10½ years ago!? :astonished:

## Description
Store the _SLAVE_ STATUS when pausing replicas for `mariadb-dump --dump-slave`;
use that (instead of a new SHOW _SLAVE_ STATUS) to resume replicas when the program ends.

## Release Notes
* Fix `mariadb-dump --dump-thread` starts replication threads that were not running

## How can this PR be tested?
* The new `multi_source.mariadb-dump_slave` tests this issue for multi-source and, by consequence, single-source replication.
* Edge cases and changed side-effects remain testable by existing methods such as `main.mysqldump`.

## Basing the PR against the correct MariaDB version
* [ ] *This is a new feature or a refactoring, and the PR is based against the `main` branch.*
* [x] *This is a bug fix, and the PR is based against the earliest _maintained_ branch in which the bug can be reproduced.*

## PR quality check
* [x] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/-/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
* [x] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

[MDEV-7611]: https://jira.mariadb.org/browse/MDEV-7611

### CODE REVIEW

File: client/mysqldump.c [hunk 1]
```diff
@@ -6250,43 +6264,29 @@ static int do_show_slave_status(MYSQL *mysql_con, int have_mariadb_gtid,
...
-
-  /* We need to check if the slave sql is stopped in the first place */
-  if (mysql_query_with_error_report(mysql_con, &slave,
-                                    multi_source ?
-                                    "SHOW ALL SLAVES STATUS" :
-                                    "SHOW SLAVE STATUS"))
-    DBUG_RETURN(1);
-
-  while ((row= mysql_fetch_row(slave)))
+  for (; n_stopped_replicas--;)
```
Comment (ParadoxV5) [inline client/mysqldump.c:6269]: To use the new list instead of SHOW SLAVE STATUS, I essentially rewrote `do_start_slave_sql`…

File: client/mysqldump.c [hunk 2]
```diff
@@ -6097,12 +6099,22 @@ static int do_stop_slave_sql(MYSQL *mysql_con)
...
+  // do_stop_slave_sql() should only be called once
+  DBUG_ASSERT(!stopped_replicas);
 
   if (mysql_query_with_error_report(mysql_con, &slave,
                                     multi_source ?
                                     "SHOW ALL SLAVES STATUS" :
                                     "SHOW SLAVE STATUS"))
     return(1);
+  stopped_replicas= my_malloc(PSI_NOT_INSTRUMENTED,
+                              slave->row_count*NAME_CHAR_LEN + 1, MYF(MY_WME));
```
Comment (ParadoxV5) [inline client/mysqldump.c:6111]: I feel that it’s wrong wasting memory on short names and already-stopped replicas, but I feel that it’s more wrong spending extra effort `realloc`ing unpredictably.

I’d like to iterate the `MYSQL_RES` an additional time before the following loop to sum up name sizes of (only) active replications.
Question: How to rewind it from `mysql_fetch_row()`s?

---

Note I didn’t check size overflow here because of [the Broken Windows Theory](https://en.wikipedia.org/wiki/Broken_windows_theory):
https://github.com/MariaDB/server/blob/6aa47fae304bd1f3efc0ebd2035bc1481ded73a3/client/mysqldump.c#L1986
https://github.com/MariaDB/server/blob/6aa47fae304bd1f3efc0ebd2035bc1481ded73a3/client/mysqldump.c#L6620

File: client/mysqldump.c [hunk 3]
```diff
@@ -6123,6 +6135,8 @@ static int do_stop_slave_sql(MYSQL *mysql_con)
           mysql_free_result(slave);
           return 1;
         }
+        strmov(stopped_replicas[n_stopped_replicas++],
+               multi_source ? row[0] : "");
```
Comment (ParadoxV5) [inline client/mysqldump.c:6139]: pedantically supporting no-multi-source on a best effort basis
(But those pre-10.0 versions were long-EOL… 😕)
https://github.com/MariaDB/server/blob/6aa47fae304bd1f3efc0ebd2035bc1481ded73a3/client/mysqldump.c#L7010-L7015

File: client/mysqldump.c [hunk 4]
```diff
@@ -6250,43 +6264,29 @@ static int do_show_slave_status(MYSQL *mysql_con, int have_mariadb_gtid,
...
-                        multi_source ? row[0] : "", row[11 + multi_source]));
-    if (row[11 + multi_source])
+    /*
+      do_start_slave_sql() should only be called
+      sometime after do_stop_slave_sql() suceeds
+    */
+    char* stopped_replica= stopped_replicas[n_stopped_replicas];
+    char query[sizeof("START SLAVE ''") + NAME_CHAR_LEN];
+    DBUG_PRINT("info", ("Connection: '%.*s'", NAME_CHAR_LEN, stopped_replica));
+    // if SLAVE SQL is running, start it anyway to warn unexpected state change
```
Comment (ParadoxV5) [inline client/mysqldump.c:6278]: The program runs SHOW SLAVE STATUS multiple times thoughout the various step functions:
* `do_show_slave_status`
* `do_stop_slave_sql`
* `do_start_slave_sql` (before this PR)

Perhaps the intent is in case replication configs change in the middle of `mariadb-dump`?

File: mysql-test/suite/multi_source/mariadb-dump_slave.test [hunk 5]
```diff
@@ -0,0 +1,42 @@
+# Interactions with --dump-slave; based on `main.rpl_mysqldump_slave`
```
Comment (ParadoxV5) [inline mysql-test/suite/multi_source/mariadb-dump_slave.test:1]: and `multi_source.gtid_slave_pos` for the setup

File: mysql-test/suite/rpl/t/rpl_mysqldump_gtid_slave_pos.test [hunk 6]
```diff
@@ -71,7 +71,6 @@ select @@global.gtid_slave_pos as "after initial slave got in sync";
 --echo # 3. A
 # Two dumps prepared to be restored in the following loop
 --exec $MYSQL_DUMP_SLAVE --dump-slave  --gtid mysql gtid_slave_pos > $MYSQLTEST_VARDIR/tmp/dump_2.sql
---source include/stop_slave.inc
```
Comment (ParadoxV5) [inline mysql-test/suite/rpl/t/rpl_mysqldump_gtid_slave_pos.test:74]: This test was aware 😄.

Comment (ParadoxV5) [inline mysql-test/suite/rpl/t/rpl_mysqldump_gtid_slave_pos.test:74]: This test was aware 😄.

**Review by bnestere — CHANGES_REQUESTED** [review]
Thanks for a nice patch, @ParadoxV5 !

Mostly some minor notes for your consideration, along with a couple test additions.

Also, to the git commit message, could you flip the solution and problem, so the problem is first?

File: client/mysqldump.c [hunk 7]
```diff
@@ -6097,12 +6099,22 @@ static int do_stop_slave_sql(MYSQL *mysql_con)
 {
   MYSQL_RES *slave;
   MYSQL_ROW row;
+  // do_stop_slave_sql() should only be called once
+  DBUG_ASSERT(!stopped_replicas);
```
Comment (ParadoxV5) [inline client/mysqldump.c:6103]: This config has a problem similar to what [MDEV-17856](https://jira.mariadb.org/browse/MDEV-17856) had:
There’re two different (sets of) entry points to `do_start_slave_sql()` that passing `stopped_replicas` to it as an argument is impractical if not impossible.
A global variable is the most it could have.

File: client/mysqldump.c [hunk 8]
```diff
@@ -1854,6 +1855,7 @@ static void free_resources()
     mysql_close(mysql);
     mysql= 0;
   }
+  my_free(stopped_replicas);
```
Comment (ParadoxV5) [inline client/mysqldump.c:1858]: Hm. Must it remain named `stopped_slaves`?

File: client/mysqldump.c (hunk 2 above)
Comment (ParadoxV5) [reply client/mysqldump.c:6111]: The intent was to use as few memory for the names as practical, yes, with a packed string array that joins variable-length names with `'\0'` separators.
The first iteration tallys the exact size up, then I `malloc`, and filling the buffer requires another iteration at the end.

However, after several stabs at the implementation, I’m losing interest as the feeling that this is overengineering grew inside me.

How about your perspective, @bnestere?

I considered making the `MYSQL_RES slave` query result public so that `do_start_slave_sql` can reuse that, but I’m concerned that the full result of `SHOW ALL SLAVES STATUS` is a massive waste of memory.
Though, is this what you mean by, “to be re-used by `do_stop_slave_sql()`”?

I plan to refactor these start/stop replica steps to use `SELECT … FROM information_schema.SLAVE_STATUS WHERE …` instead.
By moving the row and column selection to the server side, the result would only contain the names needed here, meaning my memory concern will likely be obsolete.
Unfortunately, 10.11 doesn’t have `IS.SLAVE_STATUS`…

File: mysql-test/suite/multi_source/mariadb-dump_slave.test
Comment (ParadoxV5) [inline mysql-test/suite/multi_source/mariadb-dump_slave.test:1]: This is ridiculous! The only explicit SQL statements are `START`/`STOP SLAVE SQL_THREAD` for the setup/cleanup! Are we writing SQL here or scripting?

File: client/mysqldump.c (hunk 8 above)
Comment (bnestere) [reply client/mysqldump.c]: For code-level terminology variables, yes, to stick to the existing convention (so searches, etc are easier). Though for high-level descriptions (e.g. code comments/git commit messages), I often use primary/replica terminology.

File: client/mysqldump.c (hunk 2 above)
Comment (bnestere) [reply client/mysqldump.c]: > The plan was a packed string array that joins variable-length names with '\0' separators.
The first iteration tallys the exact size up, then I malloc, and filling the buffer requires another iteration at the end.
>
> However, after several stabs at the implementation, I’m losing interest as the feeling that this is overengineering grew inside me.

Generally that seems ok by me (or the safer way would be to have another `uint` array that stores the length of each connection name, rather than `'\0'` searching).


> I considered making the MYSQL_RES slave query result public so that do_start_slave_sql can reuse that, but I’m concerned that the full result of SHOW ALL SLAVES STATUS is a massive waste of memory.
Though, is this what you mean by, “to be re-used by do_stop_slave_sql()”, or was it a different idea?

Yeah that's what I meant. Definitely not the most efficient use of memory, but not massive :innocent:. 

> I plan to, on the main branch , refactor these start/stop replica steps to use SELECT … FROM information_schema.SLAVE_STATUS WHERE … instead.

Another thought is to simply have the patch target 11.8 where we have this available. There's no specific need/request to get this into 10.11 or 11.4. There's less risk that way anyhow.

Comment (ParadoxV5) [reply client/mysqldump.c]: > > I plan to, on the main branch , refactor these start/stop replica steps to use SELECT … FROM information_schema.SLAVE_STATUS WHERE … instead.
> 
> Another thought is to simply have the patch target 11.8 where we have this available. There's no specific need/request to get this into 10.11 or 11.4. There's less risk that way anyhow.

I feel that isn’t “responsible”. We only paused 10.6 bugfixes because of the new “criticals only” policy.

Though from a unique perspective, if this behavior is well known and considered a 🐞 feature rather than a 🐛, this request would instead be a “backward-incompatible change”.
(I doubt it is?)

Comment (ParadoxV5) [reply client/mysqldump.c]: > > I considered making the MYSQL_RES slave query result public so that do_start_slave_sql can reuse that, but I’m concerned that the full result of SHOW ALL SLAVES STATUS is a massive waste of memory.
> > Though, is this what you mean by, “to be re-used by do_stop_slave_sql()”, or was it a different idea?
> 
> Yeah that's what I meant. Definitely not the most efficient use of memory, but not massive 😇.

I’ll go with this suggestion of yours.
It’s the least efficient, but it’s closest to my preferred solution:

> > I plan to, on the main branch , refactor these start/stop replica steps to use SELECT … FROM information_schema.SLAVE_STATUS WHERE … instead.

I noted this follow-up as [MDEV-37146](https://jira.mariadb.org/browse/MDEV-37146); remind me to pay off this tech debt I’m about to borrow.

File: client/mysqldump.c (hunk 3 above)
Comment (ParadoxV5) [reply client/mysqldump.c]: Oh? 🧐 [The Corp’s Engineering Policy](https://mariadb.com/engineering-policies/) says that 10.5 *Enterprise* goes EOL after 16th this month.
It should be safe to refactor away no-multi-source support.

File: client/mysqldump.c [hunk 9]
```diff
@@ -6412,32 +6414,35 @@ static int do_show_slave_status(MYSQL *mysql_con, int have_mariadb_gtid,
...
 
-  /* We need to check if the slave sql is stopped in the first place */
-  if (mysql_query_with_error_report(mysql_con, &slave,
-                                    multi_source ?
-                                    "SHOW ALL SLAVES STATUS" :
-                                    "SHOW SLAVE STATUS"))
-    DBUG_RETURN(1);
+  /*
+    do_start_slave_sql() should normally be called
+    sometime after do_stop_slave_sql() suceeds
```
Comment (bnestere) [inline client/mysqldump.c]:
```suggestion
    sometime after do_stop_slave_sql() succeeds
```

**Review by bnestere — APPROVED** [review]
Looks good, thanks @ParadoxV5 !
